packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -102,12 +102,7 @@ const processNodeUrl = (
     originalUrl !== '/' &&
     htmlPath === '/index.html'
   ) {
-    const replacer = (url: string) =>
-      path.posix.join(
-        devBase,
-        path.posix.relative(originalUrl, devBase),
-        url.slice(1)
-      )
+    const replacer = (url: string) => path.posix.join(devBase, url.slice(1))
```

Comment:
Looks like this function is added in https://github.com/vitejs/vite/pull/6419. This change seems it would fix the absolute base case `/a/b/c/`, but we might still need to use `path.posix.relative` for `base` `./`. It might need a ternary to handle both.

Reply:
I added a `joinUrlSegments` function that should handle both cases

Reply:
I found that if the base is `./`, it will be change to `/` in dev mode at [here](https://github.com/vitejs/vite/blob/main/packages/vite/src/node/config.ts#L519)
```js
  // During dev, we ignore relative base and fallback to '/'
  // For the SSR build, relative base isn't possible by means
  // of import.meta.url.
  const resolvedBase = relativeBaseShortcut
    ? !isBuild || config.build?.ssr
      ? '/'
      : './'
    : resolveBaseUrl(config.base, isBuild, logger) ?? '/'
```
The `devHtmlHook`  is only using in dev mode.
So the base url here can't be `./`

Reply:
still, it's probably safer to use the existing `joinUrlSegments` utility so that if `./` is ever allowed here in the future it will just work

Reply:
If `./` is allowed future, `joinUrlSegments `  still can not handle this two case

| caseNo | devBase | url           | joinUrlSegments Result |
| ------ | ------- | ------------- | ---------------------- |
| case1  | /a/b/c  | ./src/main.ts | /a/b/c/./src/main.ts   |
| case2  | ./      | ./src/main.ts | ././src/main.ts        |

`joinUrlSegments` can handle case1 because the result is absolute path

However, in caes2, `joinUrlSegments` will return a relative path. It doesn't work when i access the url `/aaa/bbb/ccc/ddd` or `/aaa/bbb/ccc/` which will be fallback to `/index.html` because the relative path in different url is to different file .

Therefore we still need to use `path.posix.relative` for `base` `./`

Reply:
@benmccann 
`joinUrlSegments`  will return the url like this `/a/b/c/./src/main.ts`. 
`./` is no need and it cause the test case error. 
May be we can normalize it by using `path.posix.resolve` before return?
![image](https://user-images.githubusercontent.com/20377280/202886162-25704000-dfe3-430c-8196-ca517200798e.png)

Reply:
> I found that if the base is `./`, it will be change to `/` in dev mode at [here](https://github.com/vitejs/vite/blob/main/packages/vite/src/node/config.ts#L519)

Ah nice find. I forgot this only runs in dev mode. IMO I think it's fine that we don't handle `./` as long as we put a comment above it.

I think the first commit's fix would be good then, it doesn't looks like `joinUrlSegments` handle `./` between the paths.

Reply:
you're right. thank you!